### PR TITLE
Add federated authenticator for user portal and apps

### PR DESCRIPTION
**Pull request by madurangasiriwardena**

### CODE REVIEW

**Comment by malithie** [comment]
I don't think we need to replicate all code in the basic authenticator here. For example multi attribute login etc. We don't have such requirements now. Rather than complicating the code I think it's better to add them when such requirements come. WDYT?
Let's improve this to support present requirements we have rather than generalising.

File: components/org.wso2.carbon.identity.cloud.authenticator.federated/src/main/java/org/wso2/carbon/identity/cloud/authenticator/federated/internal/IdentityCloudAuthenticatorServiceComponent.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    private static Log log = LogFactory.getLog(IdentityCloudAuthenticatorServiceComponent.class);
+
+    private static RealmService realmService;
+
+    public static RealmService getRealmService() {
+        return realmService;
+    }
+
+    protected void setRealmService(RealmService realmService) {
+        log.debug("Setting the Realm Service");
```
Comment (dinusha-dilrukshi) [inline components/org.wso2.carbon.identity.cloud.authenticator.federated/src/main/java/org/wso2/carbon/identity/cloud/authenticator/federated/internal/IdentityCloudAuthenticatorServiceComponent.java]: Check for isDebugenabled()

File: components/org.wso2.carbon.identity.cloud.authenticator.federated/src/main/java/org/wso2/carbon/identity/cloud/authenticator/federated/internal/IdentityCloudAuthenticatorServiceComponent.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        IdentityCloudAuthenticatorServiceComponent.realmService = realmService;
+    }
+
+    protected void activate(ComponentContext ctxt) {
+        try {
+            IdentityCloudFederatedAuthenticator federatedCustomAuth = new IdentityCloudFederatedAuthenticator();
+            ctxt.getBundleContext().registerService(ApplicationAuthenticator.class.getName(), federatedCustomAuth, null);
+            if (log.isDebugEnabled()) {
+                log.info("IdentityCloudFederatedAuthenticator bundle is activated");
```
Comment (dinusha-dilrukshi) [inline components/org.wso2.carbon.identity.cloud.authenticator.federated/src/main/java/org/wso2/carbon/identity/cloud/authenticator/federated/internal/IdentityCloudAuthenticatorServiceComponent.java]: Should be log.debug() not log.info()

File: components/org.wso2.carbon.identity.cloud.authenticator.federated/src/main/java/org/wso2/carbon/identity/cloud/authenticator/federated/internal/IdentityCloudAuthenticatorServiceComponent.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            }
+        } catch (Throwable e) {
+            log.error("IdentityCloudFederatedAuthenticator bundle activation Failed", e);
+        }
+    }
+
+    protected void deactivate(ComponentContext ctxt) {
+        if (log.isDebugEnabled()) {
+            log.info("IdentityCloudFederatedAuthenticator bundle is deactivated");
```
Comment (dinusha-dilrukshi) [inline components/org.wso2.carbon.identity.cloud.authenticator.federated/src/main/java/org/wso2/carbon/identity/cloud/authenticator/federated/internal/IdentityCloudAuthenticatorServiceComponent.java]: Should be log.debug()

File: components/org.wso2.carbon.identity.cloud.authenticator.federated/src/main/java/org/wso2/carbon/identity/cloud/authenticator/federated/internal/IdentityCloudAuthenticatorServiceComponent.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        }
+    }
+
+    protected void unsetRealmService(RealmService realmService) {
+        log.debug("UnSetting the Realm Service");
```
Comment (dinusha-dilrukshi) [inline components/org.wso2.carbon.identity.cloud.authenticator.federated/src/main/java/org/wso2/carbon/identity/cloud/authenticator/federated/internal/IdentityCloudAuthenticatorServiceComponent.java]: Check for isDebugenabled()

**Review by dinusha-dilrukshi — CHANGES_REQUESTED** [review]

**Comment by madurangasiriwardena** [comment]
@malithie , we have created [1] to track this improvement.

[1] https://wso2.org/jira/browse/IDNCLOUD-221

**Review by dinusha-dilrukshi — APPROVED** [review]

**Pull request merged**